extensions/vertx-web-common/runtime/src/main/java/io/quarkus/vertx/web/common/runtime/VertxWebCommonRecorder.java

```diff
@@ -0,0 +1,462 @@
+public class VertxWebCommonRecorder {
+
+    private static final Logger LOGGER = Logger.getLogger(VertxWebCommonRecorder.class.getName());
+
+    private static volatile Handler<RoutingContext> hotReplacementHandler;
+
+    private static volatile Router router;
+
+    private static volatile Runnable closeTask;
+
+    public static void setHotReplacement(Handler<RoutingContext> handler) {
+        hotReplacementHandler = handler;
+    }
+
+    public static void shutDownDevMode() {
+        closeTask.run();
+        closeTask = null;
+        router = null;
+        hotReplacementHandler = null;
+    }
+
+    public static void startServerAfterFailedStart() {
+        VertxConfiguration vertxConfiguration = new VertxConfiguration();
+        ConfigInstantiator.handleObject(vertxConfiguration);
+        VertxCommonRecorder.initialize(vertxConfiguration);
+
+        try {
+            HttpConfiguration config = new HttpConfiguration();
+            ConfigInstantiator.handleObject(config);
+
+            router = Router.router(VertxCommonRecorder.getVertx());
+            if (hotReplacementHandler != null) {
+                router.route().blockingHandler(hotReplacementHandler);
+            }
+
+            //we can't really do
+            doServerStart(VertxCommonRecorder.getVertx(), config, LaunchMode.DEVELOPMENT);
+
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public RuntimeValue<Router> initializeRouter(RuntimeValue<Vertx> vertxRuntimeValue, BeanContainer container,
+            HttpConfiguration httpConfiguration, LaunchMode launchMode,
+            boolean startVirtual, boolean startSocket) throws IOException {
+
+        Vertx vertx = vertxRuntimeValue.getValue();
+
+        if (router == null) {
+            router = Router.router(vertx);
+            if (hotReplacementHandler != null) {
+                router.route().handler(hotReplacementHandler);
+            }
+        }
+        // Make it also possible to register the route handlers programmatically
+        Event<Object> event = Arc.container().beanManager().getEvent();
+        event.select(Router.class).fire(router);
+
+        if (startVirtual) {
+            initializeVirtual(vertx);
+        }
+        if (startSocket) {
+            // Start the server
+            if (closeTask == null) {
+                doServerStart(vertx, httpConfiguration, launchMode);
+            }
+        }
+
+        return new RuntimeValue<>(router);
+    }
+
+    public void finalizeRouter(BeanContainer container, Handler<HttpServerRequest> defaultRouteHandler,
+            List<Handler<RoutingContext>> filters, LaunchMode launchMode, ShutdownContext shutdown) {
+        // install the default route at the end
+        if (defaultRouteHandler != null) {
+            //TODO: can we skip the router if no other routes?
+            Route defaultRoute = router.route();
+
+            for (Handler<RoutingContext> filter : filters) {
```

Comment:
I think the original code was wrong before because it was adding filters as routes instead of adding them as handlers to the default route. But a quick check would be appreciated.

See https://github.com/quarkusio/quarkus/blob/master/extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/VertxWebRecorder.java#L161